### PR TITLE
User schemas

### LINKED ISSUE #12: Create schema & model for User objects
Create `/Users/userModels.js`  
Using `mongoose` and `mongoose.Schema` create schema and models for User objects.  
Refer to the [Database Schema Reference](https://github.com/TTUSDC/CPCEEDWebAppBackend/wiki/Database-Schema-Reference#users) for more details.

### PR DESCRIPTION
**Pull request by ryanfaulkenberry100**
Closes #12

### CODE REVIEW

**Comment by asclines** [comment]
@ryanfaulkenberry100 As per our plans to comply with Google Style Guides and whatnot, could you change the `Users` folder to `users`?

File: Users/router.js
Comment (jmtaber129) [inline Users/router.js]: The URI for this should just be ```'/'```, since you wouldn't know the UID of a user that hasn't been created, yet.

File: Users/service.js
Comment (jmtaber129) [inline Users/service.js]: ```Student``` -> ```student```.  Local variable names should be in lowerCamelCase: https://google.github.io/styleguide/jsguide.html#naming-local-variable-names

File: Users/service.js
Comment (jmtaber129) [inline Users/service.js]: There should be no line breaks between the closing bracket of the ```if``` and the ```else if (...) {```: https://google.github.io/styleguide/jsguide.html#formatting-nonempty-blocks

File: Users/router.js
Comment (jmtaber129) [inline Users/router.js]: All comments should in with proper punctuation (in this case a period).

File: Users/service.js
Comment (jmtaber129) [inline Users/service.js]: ```StudentModel``` is not defined here.  You need to ```require()``` the model file, and use the exported models from that.

File: Users/service.js
Comment (jmtaber129) [inline Users/service.js]: Since the way in which authentication information will be passed hasn't been decided, I'd just leave checking whether it's an admin as a TODO.

File: Users/userModels.js
Comment (jmtaber129) [inline Users/userModels.js]: This file should be named ```user-models.js```: https://google.github.io/styleguide/jsguide.html#file-name

File: Users/router.js
Comment (jmtaber129) [inline Users/router.js]: I'd just call the models ```Student``` and ```Admin```.

File: Users/service.js
Comment (jmtaber129) [inline Users/service.js]: Should ```approvalStatus``` be accepted for this method?  This way, a student could potentially approve themselves when they create their account.

**Review by jmtaber129 — CHANGES_REQUESTED** [review]
First pass, will probably have additional comments later.

Make sure your indentation is two spaces, and that your indentation is consistent throughout a given block.  Tabs should NOT be used.  Change your text editor settings to emit spaces when you press the tab key.

All comments should end with proper punctuation.

I generally prefer TODOs of the form ```// TODO(<username>): Stuff that needs to be done.```.

**Comment by ryanfaulkenberry100** [comment]
Applied all suggestions, ready for further review

**Review by asclines — DISMISSED** [review]
I'm approving this because `user-models.js` looks great! And that was the assignment for this PR. the manager and router will need work now

File: users/router.js
Comment (jmtaber129) [inline users/router.js]: Since the models aren't used here, go ahead and remove the ```require(...)```

File: users/service.js
Comment (jmtaber129) [inline users/service.js]: This should be ```require('./user-models.js')``` per the filename change.

File: users/service.js
Comment (jmtaber129) [inline users/service.js]: Is this field the unique identifier (UID created by the DB), or the student's R number?  If it's the UID, this field is created and populated as either '__id' or 'id' by the database on creation.  If it's the R number, this should be json.studentId.

Additionally, ```getUid()``` is not defined.

File: users/service.js
Comment (jmtaber129) [inline users/service.js]: Nit: I'd prefer for the blank line to be above the TODO here.

File: users/service.js
Comment (jmtaber129) [inline users/service.js]: Fix indentation (4 spaces -> 2 spaces).

File: users/service.js
Comment (jmtaber129) [inline users/service.js]: Since if and else-if statements need a condition to be syntactically correct, go ahead and stub the condition (e.g. true/false) for the if and else-if.  A brief description of what the actual condition will be might also be good.

For example: ```if (/*modifying user is student*/ true) {```

File: users/service.js
Comment (jmtaber129) [inline users/service.js]: Go ahead and add an else statement, along with a TODO for error handling.

File: users/service.js
Comment (jmtaber129) [inline users/service.js]: Blank lines at the start of a function body are not allowed: https://google.github.io/styleguide/jsguide.html#formatting-vertical-whitespace

File: users/service.js
Comment (jmtaber129) [inline users/service.js]: ```User``` is not defined here.

File: users/service.js
Comment (jmtaber129) [inline users/service.js]: ```modifyUserAsSelf()``` is not defined.

File: users/router.js
Comment (jmtaber129) [inline users/router.js]: ```response.body``` -> ```response.object```

File: users/router.js
Comment (jmtaber129) [inline users/router.js]: Port should be 27017 (default port for MongoDB).  Valid ports are between 0 and 65535.

File: users/user-models.js
Comment (jmtaber129) [inline users/user-models.js:2]: You need to ```require('mongoose')``` here.

File: users/service.js
Comment (jmtaber129) [inline users/service.js]: Since the function name is ```modifyUserAsSelf()```, this comment is redundant.  Same with ```modifyUserAsAdmin()```.

**Review by jmtaber129 — CHANGES_REQUESTED** [review]
A few more nits, but there are also a few issues that are preventing your code from running (either the server won't start, or other errors occur during API calls).  For future PRs, please check that your code runs correctly before opening a pull request.

For functions that should be called, but have not been implemented, you can create a function stub (e.g. ```var foo = (...) => {}```) to ensure that the function is defined.

**Comment by asclines** [comment]
@ryanfaulkenberry100 is there any update on this?

File: users/service.js
Comment (ryanfaulkenberry100) [reply users/service.js]: It's a UID, I wasn't sure how UID generation would work. For now I'll just change it to 0.

File: users/router.js
Comment (asclines) [inline users/router.js]: This is no longer needed as it is handled in `server.js` 
Merge `master` into your branch to get the latest changes

File: users/router.js
Comment (asclines) [inline users/router.js]: The router shouldn't need this. Models should only be handled in the manager.

File: users/router.js
Comment (asclines) [inline users/router.js]: This should use callbacks as this will be asynchronous.

File: users/router.js
Comment (asclines) [inline users/router.js]: Response shouldn't have status. The router should be able to determine that,

File: users/user-models.js
Comment (asclines) [inline users/user-models.js]: I'm not sure how this works. I believe this should be `var mongoose = require('mongoose')`

**Review by asclines — CHANGES_REQUESTED** [review]
I'm not 100% sure what is going on with service but if you could look at `reports/report-manager.js` you can find some good examples there.

Filename requests:
- `router.js` -> 'user-router.js`
- 'service.js' -> 'user-manager.js`

**Comment by ryanfaulkenberry100** [comment]
@asclines @jmtaber129 

1. The code actually runs now (sorry bout that)

2. Applied all of your requested changes

Ready for further review

File: users/user-router.js
Comment (asclines) [inline users/user-router.js]: Is mongoose still used in this file?

File: users/user-manager.js
Comment (asclines) [inline users/user-manager.js]: Points is now an object of key value pairs where keys are the categories and values are the points for that category. However, I believe we can safely assume that all users start out with 0 points in each category. Therefore instead of needing that in the `reqData`, initialize it yourself like you did `approvalStatus`

File: users/user-manager.js
Comment (asclines) [inline users/user-manager.js]: should get the id from `reqData`

File: users/user-manager.js
Comment (asclines) [inline users/user-manager.js]: Put a TODO here to indicate that this needs to be hooked up to the `user-manager`

File: users/user-manager.js
Comment (asclines) [inline users/user-manager.js]: Put a TODO here to handle the case where the role is incorrect.

File: users/user-manager.js
Comment (asclines) [inline users/user-manager.js]: Put a TODO here to indicate that this needs to be hooked up to the `user-manager`

File: users/user-manager.js
Comment (asclines) [inline users/user-manager.js]: Put a TODO here to indicate that this needs to be hooked up to the `user-manager`

**Review by asclines — CHANGES_REQUESTED** [review]
A couple more changes and this will be good for first round. There are still some issues, but a couple TODOs there and we will tackle them in future iterations.

File: users/user-manager.js
Comment (jmtaber129) [inline users/user-manager.js]: Check indentation.

File: users/user-manager.js
Comment (jmtaber129) [inline users/user-manager.js]: Check indentation.

If the same variable name is used for the new Admin/Student, you can just call ```save()``` once (after the if-else block), in which case you'd only have one TODO for saving users.  See ```report-manager.js```.

File: users/user-manager.js
Comment (jmtaber129) [inline users/user-manager.js:82]: Why is a user being created here? When you make the mongoose ```find()``` call, you should just be using the model object.

File: users/user-router.js
Comment (jmtaber129) [inline users/user-router.js]: Check indentation.

File: users/user-manager.js
Comment (jmtaber129) [inline users/user-manager.js]: Once you save this document, mongoose will keep this field 0, and will create a separate field called ```_id``` for the generated UID.  I.e., this field would never be used as a UID.  There might be a way to set up mongoose to use this field for the UID, but I'd recommend removing this field and just using the default ```_id``` field.

**Review by jmtaber129 — CHANGES_REQUESTED** [review]
Now that we know that [mongoose discriminators](http://mongoosejs.com/docs/discriminators.html) are a thing, I'd like to see the models use them.  If you have any questions about how to use them, see ```report-models.js``` and ```report-manager.js``` or @ me on slack.

File: users/user-manager.js
Comment (asclines) [reply users/user-manager.js:82]: @jmtaber129 I think he is just doing this for testing

File: users/user-models.js
Comment (asclines) [reply users/user-models.js:2]: as opposed to `var mongoose = require('mongoose')`

File: users/user-manager.js
Comment (asclines) [reply users/user-manager.js]: @jmtaber129 what do you mean? I don't see how he is using this field as a UID

File: users/user-manager.js
Comment (jmtaber129) [reply users/user-manager.js:82]: Okay, that makes sense.

File: users/service.js
Comment (jmtaber129) [reply users/service.js]: Once you save this document, mongoose will keep this field 0, and will create a separate field called ```_id``` for the generated UID. I.e., this field would never be used as a UID. There might be a way to set up mongoose to use this field for the UID, but I'd recommend removing this field and just using the default ```_id``` field.

File: users/user-manager.js
Comment (jmtaber129) [reply users/user-manager.js]: I meant to add this comment to another thread (in which he mentioned it's the UID).  I've added this comment to the correct comment thread.

File: users/user-manager.js
Comment (asclines) [reply users/user-manager.js:82]: But it is good to point out anyways so it may be removed in the future.

**Review by asclines — APPROVED** [review]

**Comment by asclines** [comment]
@ryanfaulkenberry100 Functionally this code looks correct, and since it does accomplish the tasks it was supposed to, I'm approving this. However @jmtaber129 is probably going to come through here soon and mark all the style-guide violations (of which there are quite a few). Let's just get this code in for now. A pull-request will be appearing in the future that will handle a full on linter setup to streamline style guide conformity. Any previously committed violations will be handled then.

File: users/user-models.js
Comment (jmtaber129) [inline users/user-models.js]: I would make the ```discriminatorKey``` the ```role``` field, since that's the field that would normally show whether the user is a student or admin.  Also, having both a ```role``` and ```type``` field would be duplicating the role information.

File: users/user-manager.js
Comment (jmtaber129) [inline users/user-manager.js]: Nit: Missing period.

File: users/user-models.js
Comment (jmtaber129) [inline users/user-models.js:10]: The information stored in this field will already be stored in the field specified as the discriminator key, so the ```role``` field can be removed from the schema.

**Review by jmtaber129 — APPROVED** [review]
There are a couple changes I'd like to see before merging, but other than those, LGTM.  Like @asclines mentioned, there are still a few style violations, but I'm fine with those being fixed in a future linter PR.

**Pull request merged**
